PyAssembler/Tutorial1.py

```diff
@@ -0,0 +1,170 @@
+   }
+   return switcher.get(opcode, False)
+
+def switch_register(reg_idx):
+   switcher = {
+      'r0': "000",
+      'r1': "001",
+      'r2': "010",
+      'r3': "011",
+      'r4': "100",
+      'r5': "101",
+      'r6': "110",
+      'r7': "111"
+   }
+   return switcher.get(reg_idx, False)
+
+def switch_addressing_mode(second_arg):
+   ans = ['10', second_arg]
+   if switch_register(second_arg):
+      ans[0] = '00'
+   elif second_arg.isnumeric():
+       ans[0] = '01'
+   elif second_arg[0] == '[' and second_arg[-1] == ']':
+       ans[0] = '11'
+       ans[1] = second_arg[1:-1]
+   return ans
+
+def update_sym_tbl(label):
+   global current_address, sym_tbl
+   sym_tbl.update({label: current_address})
+
+def first_step(assembly_file):
+   global current_address, sym_tbl
+   with open(assembly_file,'r') as origin_file:
+      all_lines = origin_file.readlines()
+      for index, line in enumerate(all_lines):
+         if line == '\n' or line[0] == ';': continue
+         all_lines[index] = translate_line(line[:-1] if line[-1] is '\n' else line)                                  # remove \n
+   with open('semi_output.txt', 'w') as outfile:
+      for line in all_lines:
+         outfile.write(line + '\n')                                  # TODO remove \n before submit
+
+def second_step(outfile_name):
+   global current_address, sym_tbl
+   # read almost_binary_line
+   with open('semi_output.txt','r') as intermidiate_file:
+      all_lines = intermidiate_file.readlines()
+      for index, line in enumerate(all_lines):
+         all_lines[index] = fix_line(line[:-1] if line[-1] is '\n' else line)                                  # remove \n
+   with open('cool_' + outfile_name, 'w') as outfile:
+      for line in all_lines:
+         outfile.write(line + '\n')
+   with open(outfile_name, 'w') as outfile:
+      for line in all_lines:
+         outfile.write(line)
+   print("Assembling Done.")
+
+def validate_expression(code_line):
+   tokens = code_line.split(' ')
+   print(tokens)
+
+def check_label(words):
+   if words[0][-1] == ':':
+      update_sym_tbl(words[0][:-1])
+      return words[1:]
+   return words
+
+def translate_line(ass_line):
+   global current_address, sym_tbl
+   translated_word1 = '0000000000000000'
+   translated_word2 = '0000000000000000'
+   words = ass_line.split(' ')
+   words = check_label(words)
+   bin_opcode = switch_opcode(words[0])
+   if not bin_opcode:
+      return section_handler(words)
+   translated_word1 = bin_opcode + translated_word1[4:]                                # update opcode bits
+   line_type = len(words)
+   if(line_type == 1):                                # nop
+      current_address+=2
+      return translated_word1
+   addr_mode = switch_addressing_mode(words[-1])
+   translated_word1 = translated_word1[:7] + addr_mode[0] + translated_word1[9:]       # update addr_mode bits
+   if(line_type == 2):                                # not clr inc dec jmp jne jz sections

```

Comment:
Notice that some of the opcodes require one operand and others requires two operands, and cross using them should not work (for example, your code should not accept `mov r1`)